### PR TITLE
perf(tx-pool): reuse write lock to insert txs batch

### PR DESCRIPTION
**Pull request by hai-rise**
Currently, each transaction insertion competes for a write lock of the pool. This can be very congested for performance chains that expect 100k+ new transactions every second. Even more challenging if the block time is small -- canonical state changes also compete for the write lock several times a second.

Ideally, the RPC server can buffer transactions to insert in a batch when the pool write lock is unavailable, instead of competing for the write lock (and block anyway) for every insertion request. 

I guess this PR is a baby step towards that 😅, to start reusing a single write lock in `PoolInner::add_transactions`.

### CODE REVIEW

File: crates/transaction-pool/src/pool/mod.rs [hunk 1]
```diff
@@ -896,7 +907,7 @@ where
     ///
     /// If some of the transactions are blob transactions, they are also removed from the blob
     /// store.
-    pub(crate) fn discard_worst(&self) -> HashSet<TxHash> {
```
Comment (hai-rise) [inline crates/transaction-pool/src/pool/mod.rs:899]: Unrelated change but this sets up for the micro-optimization mentioned above. And it's easier to optimize this whole thing if we can limit the `pub(crate)` scope.

File: crates/transaction-pool/src/pool/mod.rs [hunk 2]
```diff
@@ -541,15 +542,25 @@ where
...
+        let mut added = {
+            let mut pool = self.pool.write();
+            transactions
+                .into_iter()
+                .map(|tx| self.add_transaction(&mut pool, origin, tx))
+                .collect::<Vec<_>>()
+        };
 
         // If at least one transaction was added successfully, then we enforce the pool size limits.
+        // A micro-optimization is to reuse the `pool` write lock in the `added` scope above.
```
Comment (mattsse) [inline crates/transaction-pool/src/pool/mod.rs:563]: this comment seems a bit out of place here?

can we move this into the scope above?

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks, I think this makes sense, 
even inserting a large number of transactions shouldn't block too long, can't remember what the numbers are but should only be in the µs

pedantic doc nit

**Comment by mattsse** [comment]
for the rpc part, this could be a dedicated task that receives all new txs, batches them and submits them in batches

File: crates/transaction-pool/src/pool/mod.rs [hunk 3]
```diff
@@ -541,15 +542,25 @@ where
     }
 
     /// Adds all transactions in the iterator to the pool, returning a list of results.
+    ///
+    /// Note: A large batch may lock the pool for a long time that blocks important operations
+    /// like updating the pool on canonical state changes. The caller should consider having
+    /// a max batch size to balance transaction insertions with other updates.
```
[the remark this answers is not in the document]
Comment (hai-rise) [reply crates/transaction-pool/src/pool/mod.rs:532]: Yes, it seems safe for now but eventually we should have:
- Performance chains building blocks in small intervals (Arbitrum already ~250ms, many are going lower!) and the block builder may be blocked by the mempool removing transactions from canonical state changes, then they would prioritize the lock for state changes more than adding transactions (that can still buffer for the next batch).
- More advanced mempool choices (like DAG-based for parallel execution) that have higher complexity for transaction insertion!

File: crates/transaction-pool/src/pool/mod.rs (hunk 2 above)
Comment (hai-rise) [reply crates/transaction-pool/src/pool/mod.rs]: Unlazied myself and implemented it right in this PR now 😅.

File: crates/transaction-pool/src/pool/mod.rs [hunk 4]
```diff
@@ -1298,23 +1304,18 @@ mod tests {
...
-                    TransactionOrigin::External,
-                    TransactionValidationOutcome::Valid {
-                        balance: U256::from(1_000),
-                        state_nonce: 0,
-                        transaction: ValidTransaction::ValidWithSidecar {
-                            transaction: tx,
-                            sidecar: sidecar.clone(),
-                        },
-                        propagate: true,
+            test_pool.add_transactions(
```
Comment (hai-rise) [inline crates/transaction-pool/src/pool/mod.rs:1307]: I changed the test to use the `pub` function to ensure the pool always enforces the size in "real usage". The previous tested two private functions used in conjunction which wasn't as guaranteed.

**Comment by hai-rise** [comment]
> for the rpc part, this could be a dedicated task that receives all new txs, batches them and submits them in batches

I cannot work on this immediately so created #12811. Maybe someone familiar with the RPC codebase can kill it better!

**Comment by hai-rise** [comment]
@mattsse Kind reminder that this is available for review again 🙏.

File: crates/transaction-pool/src/pool/mod.rs [hunk 5]
```diff
@@ -525,33 +526,52 @@ where
...
-        // If at least one transaction was added successfully, then we enforce the pool size limits.
-        let discarded =
-            if added.iter().any(Result::is_ok) { self.discard_worst() } else { Default::default() };
+            (added, discarded)
+        };
 
-        if discarded.is_empty() {
-            return added
-        }
+        if !discarded.is_empty() {
```
Comment (mattsse) [inline crates/transaction-pool/src/pool/mod.rs:556]: why this change?

File: crates/transaction-pool/src/pool/mod.rs [hunk 6]
```diff
@@ -876,20 +896,6 @@ where
...
 
-    /// Enforces the size limits of pool and returns the discarded transactions if violated.
-    ///
-    /// If some of the transactions are blob transactions, they are also removed from the blob
-    /// store.
-    pub(crate) fn discard_worst(&self) -> HashSet<TxHash> {
-        let discarded = self.pool.write().discard_worst();
-
-        // delete any blobs associated with discarded blob transactions
-        self.delete_discarded_blobs(discarded.iter());
```
Comment (mattsse) [inline crates/transaction-pool/src/pool/mod.rs:887]: this logic has been removed now, right?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/transaction-pool/src/pool/mod.rs (hunk 5 above)
Comment (hai-rise) [reply crates/transaction-pool/src/pool/mod.rs:556]: This is only refactoring that doesn't change semantics. 

Previously:
- If `discarded` is empty then early return `added`
- Else make some changes to `added`
- Return `added` at the end

Now:
- If `discarded` is not empty then make changes to `added`
- Return `added` at the end

It also groups discard handling in a self-explanatory `if` scope 🙏.

File: crates/transaction-pool/src/pool/mod.rs (hunk 6 above)
Comment (hai-rise) [reply crates/transaction-pool/src/pool/mod.rs:887]: It's moved to the only call site (`add_transactions`) now, in the `if !discarded.is_empty()` check.

**Review by mattsse — APPROVED** [review]
gotcha, thanks for walking me through this, 

makes sense, the only location where have larger batches is in p2p and inserting + discarding should be quite fast even for multiple txs

**Comment by mattsse** [comment]
if you want me to take care of conflicts, please open prs from non-org forks :)

pending conflicts

**Comment by hai-rise** [comment]
@mattsse No worries, let us fix all the conflicts (done for this PR) 🙏.
